appengine/pond_storage/rank.py

```diff
@@ -15,72 +15,165 @@
+  """Stores match request information."""
+  entry_keys = ndb.KeyProperty(repeated=True, kind=LeaderboardEntry)
   creation_time = ndb.DateTimeProperty(auto_now_add=True)
-  # Expiry time.
-  expiry_time = ndb.ComputedProperty(lambda self: self.creation_time + datetime.timedelta(hours=1))
+  expiry_time = ndb.ComputedProperty(
+      lambda self: self.creation_time + datetime.timedelta(minutes=1))
 
   @classmethod
-  def contains_duck(cls, duck_key):
-    return cls.query(cls.ducks == duck_key).count(limit=1)
-
-# Cleanup expired match requests.
-now = datetime.datetime.now()
-expired_requests = MatchRequest.query(MatchRequest.expiry_time < now)
-for request in expired_requests:
-  request.key.delete()
-
-# Create a new match request.
-# 1. Choose duck with highest instability not already in a match request.
-target_duck = None
-entries_query = LeaderboardEntry.query().order(-LeaderboardEntry.instability)
-for entry in entries_query:
-  # Check if it exists in a current match request
-  if not MatchRequest.contains_duck(entry.duck_key):
-    target_duck = entry.duck_key.get()
-    if target_duck:
+  def contains_entry(cls, entry_key):
+    return cls.query(cls.entry_keys == entry_key).count(limit=1)
+
+def create_ranking_match():
+  """Creates a new ranking match if a valid match can be created.
+
+  Chooses the leaderboard entry with the highest instability no already being
+  re-evaluated in a ranking match and finds three more entries relative to it
+  are also not involved in a pending match. If four valid entries can be found,
+  the entry keys are stored in a MatchRequest (for verification of result) and
+  the information needed to evaluate the matches is returned, otherwise returns
+  None.
+
+  Returns:
+    A dictionary containing the keys:
+        match_key: The urlsafe key for the ranking match created
+        duck_list: A list of dictionaries for each entry chosen containing:
+            entry_key: The urlsafe key of the chosen LeaderboardEntry
+            js: The javascript code for the duck associated with that entry
+    Or None if no valid match can be created.
+  """
+  # TODO handle dummy ducks.
+  # 1. Find the entry with the highest instability not in a match request.
+  unstable_leaderboard_entry = None
+  unstable_entries_query = (
+      LeaderboardEntry.query().order(-LeaderboardEntry.instability))
+  for entry in unstable_entries_query:
+    # Check if it exists in a current match request (ignoring dummy ducks)
+    if (not MatchRequest.contains_entry(entry.key) and
+        entry.duck_key == get_dummy_duck_key()):
+      unstable_leaderboard_entry = entry
       break
-if not target_duck:
-  print("Status: 204 No match requests available")
-else:
-  # 2. Choose up to 3 opponents based on chosen duck.
-  # TODO choose opponents based on chosen duck's instability.
-  duck_keys = [target_duck.key]
+  if not unstable_leaderboard_entry:
+    logging.info('No ducks available for a match.')
+    return None
```

Comment:
Should we send some kind of info to the client in this case?

Reply:
I don't think it matters to the client for which reason there isn't a match available.